fix(recording): add presentation minimized events + screenshare as content fixes

germanocaumo opened the pull request:
### What does this PR do?

- Include the info when presentation is minimized by presenter with push layout to layout.xml so it can be used in playback:
```
<?xml version="1.0" encoding="UTF-8"?>
<recording id="layout_swap_events">
  <event timestamp="0.0" show_screenshare="false" show_presentation="true"/>
  <event timestamp="10.8" show_screenshare="true" show_presentation="true"/>
  <event timestamp="22.5" show_screenshare="false" show_presentation="true"/>
  <event timestamp="30.4" show_screenshare="true" show_presentation="true"/>
  <event timestamp="43.0" show_screenshare="true" show_presentation="false"/>
  <event timestamp="45.6" show_screenshare="true" show_presentation="true"/>
  <event timestamp="47.1" show_screenshare="true" show_presentation="false"/>
  <event timestamp="48.5" show_screenshare="true" show_presentation="true"/>
</recording>
```
- Only consider swap/layout events inside recordings.
- Move logic to extract events to events.rb so it can be reused by other formats
- Add slides swaps to shapes.svg when quick swap occur so playback can show the slides and thumbnails when screenshare is on for the entire recording.
  - In this case, previously only the screenshare was shown.
  
To be used with https://github.com/bigbluebutton/bbb-playback/pull/331

antobinary commented:
Thank you!

kepstin reviewed (changes requested):
I think this is implemented at the wrong level in the code, and this is going to cause problems for the dynamic layout work that I'm designing for the video format right now.

The important thing is that the EDL data structures are intended to represent the times when a video is being shared vs. when it is not being shared. The *placement* of those areas within the video is decided separately.

Keep in mind that the "SetScreenshareAsContentEvent" is actually supposed to *swap* the positions of the presentation area and screenshare, and in some layouts both areas might be shown at the same time. So the deskshare video must still be added to the EDL regardless of whether the screenshare as content flag is set or not.

At this point I'm not totally sure about how to actually represent this data. I was planning to proceed with having the swap events tracked separately (i.e. in the format generated by the `get_screenshare_as_content_events`) and then using that in the video renderer to select between one of two different layouts.

kepstin commented:
FWIW, I do not believe that this magnitude of change is suitable for inclusion in 3.0.

An acceptable change would be fixing the `get_screenshare_as_content_events` method to correctly take into account recording start/stop times.

germanocaumo commented:
@kepstin Hmm ok I see, didn't think about this dynamic layout possibility.
Ok I can fix `get_screenshare_as_content_events` instead (was doing that initially already), and you do the work in the video render script?

kepstin commented:
Yes - I'm hopeful to have something ready on the video renderer fairly soon, although I will be submitting that change only for 3.1 (it depends on some changes made in https://github.com/bigbluebutton/bigbluebutton/pull/23887 )

germanocaumo committed:
fix(recording): screenshare quick swap thumbnails in presentation playback

- Only consider swap events inside recordings.
- Add slides swaps to shapes.svg when quick swap occur so playback can show the slides and thumbnails when screenshare is on for the entire recording.
-- In this case, previously only the screenshare was shown.

germanocaumo commented:
@kepstin can you review again?

kepstin commented on record-and-playback/core/lib/recordandplayback/generators/events.rb:
```diff
@@ -1180,8 +1180,7 @@ def self.get_screenshare_as_content_events(events)
       end
-
-      screenshare_content_events
+      screenshare_content_events.sort_by! { |event| event[:timestamp] }
```
Is this necessary? The events in the events.xml are already in sorted order ([this is done in the archive step](https://github.com/bigbluebutton/bigbluebutton/blob/v3.0.x-release/record-and-playback/core/lib/recordandplayback/events_archiver.rb#L345-L358)).

kepstin commented on record-and-playback/presentation/scripts/publish/presentation.rb:
```diff
@@ -957,6 +957,17 @@ def process_presentation(package_dir)
       end
 
+    when 'SetScreenshareAsContentEvent'
+      if @presentation_props['include_deskshare']
```
Please do
```ruby
next unless @presentation_props['include_deskshare']
```
to reduce indentation.

kepstin commented on record-and-playback/presentation/scripts/publish/presentation.rb:
```diff
@@ -957,6 +957,17 @@ def process_presentation(package_dir)
+    when 'SetScreenshareAsContentEvent'
+      if @presentation_props['include_deskshare']
+        screenshareAsContent = event.at_xpath('screenshareAsContent')&.text == "true"
+        if !screenshareAsContent
```
Please switch the if statement around to check the positive case first.

kepstin commented on record-and-playback/presentation/scripts/publish/presentation.rb:
```diff
@@ -957,6 +957,17 @@ def process_presentation(package_dir)
+      if @presentation_props['include_deskshare']
+        screenshareAsContent = event.at_xpath('screenshareAsContent')&.text == "true"
+        if !screenshareAsContent
+          deskshare = false
```
Can you confirm what setting `deskshare` to true/false does?

It looks like it selects whether the slide thumbnail or screensharing icon is displayed in the slide list at the bottom of the screen, is that correct?

germanocaumo replied:
Yes, it generates entries in shapes.svg file that are used to show the thumbnails and slide change logic

kepstin commented on record-and-playback/presentation/scripts/publish/presentation.rb:
```diff
@@ -1313,15 +1324,27 @@ def process_swap_events(events)
+          timestamp: (translate_timestamp(event[:timestamp].to_f) / 1000).round(1),
+          show_screenshare: event[:screenshareAsContent],
+        )
+      end
```
Rather than implement this complex (and slightly incorrect) loop to match up recording events and swap events in the presentation format, I'd like to see code to do this added to the `generators/events.rb` file in the `get_screenshare_as_content_events` method. That way it can be re-used by multiple recording formats.

The method should also accept start/stop times so it can be used with segmented recording processing. (See how `get_chat_events` works for reference.)

Rather than having nested loops, I recommend doing a single iteration through both swap events and recording events, keeping track of the current swap state, recording state, and timestamp offset.

Note that the `translate_timestamp` method iterates through all the recording events each time it is called, making this method quadratic over the recording events. It should generally be avoided.

germanocaumo committed:
refactor(recording): apply suggestions from review

kepstin committed:
Simplify recording state tracking, fix start_time handling

The recording state tracking was unnecessarily complicated in order to
handle the case where a recording start (or stop) was received while the
recording was already started (or stopped). None of the other recording
bothers handling this, and the conditions can be made easier to read by
removing this logic.

To simplify the tracking of last screenshare as content, I'm not reading
it directly from the screenshare_content_events array instead of trying
to keep a separate variable in sync.

Code has been added to handle the special case of the start_time
(recording segment start timestamp) happening after a recording start
event and screenshare as content event have been received. In this case,
it needs to emit an event corresponding to the current screenshare as
content state at the 0 timestamp.

kepstin commented:
I've pushed an update to this with the following changes:

The recording state tracking was unnecessarily complicated in order to handle the case where a recording start (or stop) was received while the recording was already started (or stopped). None of the other recording code bothers handling this, and the conditions can be made easier to read by removing this logic.

To simplify the tracking of last screenshare as content, I'm reading it directly from the `screenshare_content_events` array instead of trying to keep a separate variable in sync.

Code has been added to handle the special case of the `start_time` (recording segment start timestamp) happening after a recording start event and screenshare as content event have been received. In this case, it needs to emit an event corresponding to the current screenshare as content state at the 0 timestamp.

@germanocaumo I'd appreciate you taking a look through my modified code and testing it to make sure it works as expected.

germanocaumo commented:
@kepstin sorry I'm on vacation, will only be able to test it properly when back 10th

kepstin committed:
Fix offset calculation for segmented recordings

germanocaumo commented on record-and-playback/core/lib/recordandplayback/generators/events.rb:
```diff
@@ -1180,37 +1189,57 @@ def self.get_screenshare_as_content_events(events, start_time, end_time)
       record = events.at_xpath('/recording/event[@eventname="RecordStatusEvent"]').nil?
 
       screenshare_content_events = []
-      last_screenshare_as_content = nil
```
Missing screenshare_as_content initialization here?

germanocaumo commented:
@kepstin It looks good, only seems to be missing the screenshare_as_content variable initialization.

germanocaumo committed:
feat(recording): add presentation minize events to layout.xml

- Include the info when presentation is minimized by presenter with push layout on:

<?xml version="1.0" encoding="UTF-8"?>
<recording id="layout_swap_events">
  <event timestamp="0.0" show_screenshare="false"/>
  <event timestamp="6.0" show_screenshare="true"/>
  <event timestamp="12.7" show_screenshare="false"/>
  <event timestamp="19.7" show_presentation="true"/>
  <event timestamp="21.6" show_presentation="false"/>
  <event timestamp="24.8" show_presentation="true"/>
  <event timestamp="29.9" show_presentation="false"/>
</recording>

kepstin commented:
In the time since this was originally written, I ended up writing basically the same code to do these events for the video processing here:

https://github.com/bigbluebutton/bigbluebutton/blob/26e08d8b9dd315b424abd8129ab541f8e8ddac5a/record-and-playback/core/lib/recordandplayback/generators/events.rb#L630-L694

It might be possible to use this code from the presentation processing scripts rather than implement another method which does basically the same thing.

I think the code would look more or less like this:
```ruby
layout_edl = BigBlueButton::Events.create_layout_edl(events_xml)
layout_edl = BigBlueButton::Events.edl_match_recording_marks_video(layout_edl, events_xml, start_time, stop_time)
```

Then `layout_edl` will contain a structure like this:

```ruby
[
  {
    timestamp: 0
    areas: {},
    conditions: {
      screenshare_as_content: true,
      presentation_is_open: true,
    }
  },
  {
    timestamp: 720000,
    areas: {},
    conditions: {
      screenshare_as_content: false,
      presentation_is_open: true,
    },
    …
  },
]
```

which should be straightforwards to map to the desired `layout.xml` structure. Note that timestamp is in milliseconds.

germanocaumo committed:
refactor(recording): re-use create_layout_edl to build layout.xml
Refactored the layout event building to use the new create_layout_edl.
Xml generated:

<?xml version="1.0" encoding="UTF-8"?>
<recording id="layout_swap_events">
  <event timestamp="0.0" show_screenshare="false" show_presentation="true"/>
  <event timestamp="10.8" show_screenshare="true" show_presentation="true"/>
  <event timestamp="22.5" show_screenshare="false" show_presentation="true"/>
  <event timestamp="30.4" show_screenshare="true" show_presentation="true"/>
  <event timestamp="43.0" show_screenshare="true" show_presentation="false"/>
  <event timestamp="45.6" show_screenshare="true" show_presentation="true"/>
  <event timestamp="47.1" show_screenshare="true" show_presentation="false"/>
  <event timestamp="48.5" show_screenshare="true" show_presentation="true"/>
</recording>

kepstin commented on record-and-playback/core/lib/recordandplayback/generators/events.rb:
```diff
@@ -1339,21 +1339,32 @@ def self.screenshare_has_audio?(events, deskshare_dir)
+      layout_edl = BigBlueButton::Events.create_layout_edl(events)
+      layout_edl = BigBlueButton::Events.edl_match_recording_marks_video(layout_edl, events, start_time, end_time)
+
+      layout_edl.each_with_object([]) do |entry, events|
```
To avoid confusion, please use a different variable name instead of `events` here. This shadows the `events` variable which is a parameter to the `get_layout_events` method.

germanocaumo committed:
chore(recording): rename variable
